### ISSUE TITLE
[Enhancement] Improved public API for IDownstreamWebApi 

### ISSUE THREAD

**Issue by jmprieur** [AzureAD/microsoft-identity-web#537]
> Thanks for your suggestion, @NikolaosWakem, and for being an early adopter.
> Would you see value in having other generic methods (for instance CallWebApiForUserAsync?)

Thanks for getting back to me... I just use the following... 

```
public static class MicrosoftIdentityWebExtensions
    {
        public static async Task<TOutput> GetWebApiForUserAsync<TOutput>(this IDownstreamWebApi downstreamWebApi, string serviceName, string relativePath)
        {
            return await downstreamWebApi.CallWebApiForUserAsync<TOutput>(serviceName, options =>
            {
                options.HttpMethod = HttpMethod.Get;
                options.RelativePath = relativePath;
            });
        }

        public static async Task PostWebApiForUserAsync(this IDownstreamWebApi downstreamWebApi, string serviceName, string relativePath)
        {
            await downstreamWebApi.CallWebApiForUserAsync(serviceName, options =>
            {
                options.HttpMethod = HttpMethod.Post;
                options.RelativePath = relativePath;
            });
        }

        public static async Task<TOutput> PostWebApiForUserAsync<TOutput, TInput>(this IDownstreamWebApi downstreamWebApi, string serviceName, string relativePath, TInput data)
        {
            return await downstreamWebApi.CallWebApiForUserAsync<TInput, TOutput>(serviceName, data, options =>
            {
                options.HttpMethod = HttpMethod.Post;
                options.RelativePath = relativePath;
            });
        }

        public static async Task PutWebApiForUserAsync<TInput>(this IDownstreamWebApi downstreamWebApi, string serviceName, string relativePath, TInput data)
        {
            await downstreamWebApi.CallWebApiForUserAsync(serviceName, data, options =>
            {
                options.HttpMethod = HttpMethod.Put;
                options.RelativePath = relativePath;
            });
        }

        public static async Task<TOutput> PutWebApiForUserAsync<TOutput, TInput>(this IDownstreamWebApi downstreamWebApi, string serviceName, string relativePath, TInput data)
        {
            return await downstreamWebApi.CallWebApiForUserAsync<TInput, TOutput>(serviceName, data, options =>
            {
                options.HttpMethod = HttpMethod.Put;
                options.RelativePath = relativePath;
            });
        }

        private static async Task<TOutput> CallWebApiForUserAsync<TOutput>(this IDownstreamWebApi downstreamWebApi, string serviceName, Action<DownstreamWebApiOptions> downstreamWebApiOptionsOverride = null, ClaimsPrincipal user = null)
        {
            var response = await downstreamWebApi.CallWebApiForUserAsync(
                serviceName,
                downstreamWebApiOptionsOverride,
                user,
                null
                ).ConfigureAwait(false);

            var content = await response.Content.ReadAsStringAsync().ConfigureAwait(false);

            if (!response.IsSuccessStatusCode)
                throw new HttpRequestException($"{(int)response.StatusCode} {response.StatusCode} {content}");

            return !string.IsNullOrWhiteSpace(content) ? JsonSerializer.Deserialize<TOutput>(content, new JsonSerializerOptions { PropertyNameCaseInsensitive = true }) : default;
        }

        private static async Task CallWebApiForUserAsync<TInput>(this IDownstreamWebApi downstreamWebApi, string serviceName, TInput input, Action<DownstreamWebApiOptions> downstreamWebApiOptionsOverride = null, ClaimsPrincipal user = null)
        {
            var response = await downstreamWebApi.CallWebApiForUserAsync(
                serviceName,
                downstreamWebApiOptionsOverride,
                user,
                new StringContent(JsonSerializer.Serialize(input), Encoding.UTF8, "application/json")
                ).ConfigureAwait(false);

            if (!response.IsSuccessStatusCode)
                throw new HttpRequestException($"{(int)response.StatusCode} {response.StatusCode} {await response.Content.ReadAsStringAsync().ConfigureAwait(false)}");
        }

        private static async Task<TOutput> CallWebApiForUserAsync<TInput, TOutput>(this IDownstreamWebApi downstreamWebApi, string serviceName, TInput input, Action<DownstreamWebApiOptions> downstreamWebApiOptionsOverride = null, ClaimsPrincipal user = null)
        {
            var response = await downstreamWebApi.CallWebApiForUserAsync(
                serviceName,
                downstreamWebApiOptionsOverride,
                user,
                new StringContent(JsonSerializer.Serialize(input), Encoding.UTF8, "application/json")
                ).ConfigureAwait(false);

            var content = await response.Content.ReadAsStringAsync().ConfigureAwait(false);

            if (!response.IsSuccessStatusCode)
                throw new HttpRequestException($"{(int)response.StatusCode} {response.StatusCode} {content}");

            return !string.IsNullOrWhiteSpace(content) ? JsonSerializer.Deserialize<TOutput>(content, new JsonSerializerOptions { PropertyNameCaseInsensitive = true }) : default;
        }

        public static bool IsMsalUiRequiredException(this Exception exception)
        {
            if (exception is Microsoft.Identity.Client.MsalUiRequiredException)
                return true;
            else if (exception.InnerException != null)
                return IsMsalUiRequiredException(exception.InnerException);
            else
                return false;
        }
    }
```

_Originally posted by @NikolaosWakem in https://github.com/AzureAD/microsoft-identity-web/issues/503#issuecomment-684073768_

**Comment by bgavrilMS** [comment]
Just as a thought, C# 8 introduced "default implementations in interfaces" which allows you to add new methods to interfaces without breaking backwards compat. So if these methods should be part of the interface, consider using this mechanism. If however these methods are more of a convenience API, or you want to hide them under a different namespace, you can use extensions.

https://devblogs.microsoft.com/dotnet/default-implementations-in-interfaces/

**Comment by jennyf19** [comment]
Included in [1.2.0 release](https://github.com/AzureAD/microsoft-identity-web/releases/tag/1.2.0).
